code/workspaces/web-app/src/components/stacks/StackCard.js

```diff
@@ -105,12 +111,18 @@ const StackCard = ({ classes, stack, openStack, deleteStack, editStack, restartS
   const expireStacks = getFeatureFlags().expireUnusedNotebooks;
 
   const generateWarningMessage = () => {
-    if (expireStacks && stack.accessTime) {
-      const { accessTimeWarning } = expireStacks;
+    if (expireStacks && stack.accessTime && stack.type) {
+      const { accessTimeWarning, inUseTimeWarning } = expireStacks;
       const daysSinceAccess = daysSinceCreation(stack.accessTime);
+      const hoursSinceAccess = hoursSinceCreation(stack.accessTime);
       const warn = (daysSinceAccess > accessTimeWarning);
-
-      if (warn) {
+      const inUseWarn = (hoursSinceAccess < inUseTimeWarning) && (stackTypes.RSTUDIO === stack.type);
+      if (!warn) {
+        if (inUseWarn) {
```

Comment:
I agree it does look a bit odd - but I originally did it your way and Intellij suggested doing the other - I decided to go with Intellij because the two warnings should be mutually exclusive since a recently opened notebook is not also one that you want to expire (or vice versa).  It is possible to set the thresholds for showing the two messages so that both messages will pass the double 'if' version, but I decided the nested 'if' gave the more common sense result.  So it just ensures one warning will appear.

Reply:
It'll still never show both warnings if you return one though - in particular the `else` block is sort of unneeded. Generally I think code that ends up less indented is clearer.